### ISSUE TITLE
Command-substitution variables clobber global variables.

### ISSUE THREAD

**Issue by 0ion9** [fish-shell/fish-shell#6480]
```fish
set v (seq 10)
echo 'Before: v = '"$v"
set foobar (for v in $v; echo $v; end)
echo "After: v = $v"
echo 'Set sez:'
set -S v
```

Output:
```
Before: v = 1 2 3 4 5 6 7 8 9 10
After: v = 10
Set sez:
$v: not set in local scope
$v: set in global scope, unexported, with 1 elements
$v[1]: length=2 value=|10|
$v: not set in universal scope
```

Okay, maybe 'correct behaviour' is a little ambiguous here. Maybe the user wants to know the inner value of v?

But here's the real problem:

```fish
set v (seq 10); echo 'Before: v = '"$v";function f; set foobar (for v in $v; echo $v; end); end; f; echo "After: v = $v";echo 'Set sez:'; set -S v
```

In other words, if you set a variable in your shell and then call a function which uses a variable of the same name in a subshell, your variable gets clobbered.  That means **if you want to ensure your variables stay intact, you cannot use the most convenient short variable names** (a,b, ..), since any function might use a subshell that happens to clobber them.

This is subshell specific. 'v' is not clobbered in the following case:

```fish
set v (seq 10); echo 'Before: v = '"$v";function f; for v in $v; echo -n "$v "; end; end; f; echo "After: v = $v";echo 'Set sez:'; set -S v
```

It is not specific to 'for' loops. 'while' loops in a subshell clobber in a different way:

```fish
set v (seq 10); echo 'Before: v = '"$v";function f; set foobar (printf %s\n $v | while read v; echo $v; end); echo $foobar; end; f; echo "After: v = $v";echo 'Set sez:'; set -S v
```

The result of the above is that, instead of v being set to 10, it is set to nothing (ie `set -g v`)

Two more cases. No loop, just set:

```fish
set v (seq 10); echo 'Before: v = '"$v";function f; set foobar (set v (string join '' $v)); end; f; echo "After: v = $v";echo 'Set sez:'; set -S v
```
(v's value is now `12345678910`)

No loop, just set -l
```fish
set v (seq 10); echo 'Before: v = '"$v";function f; set foobar (set -l v (string join '' $v)); end; f; echo "After: v = $v";echo 'Set sez:'; set -S v
```
This final case produces 'correct' results (v is not clobbered)

**Expected behaviour**:

 At least within a function, a subshell should not be able to damage variables in the global scope  without explicitly stating that it is intended (`set -g`).  In the cases of `for VARNAME` and `while VARNAME` within a subshell, VARNAME should either be set locally to the subshell, or locally to the function.

`fish --version`: fish, version 3.0.2-2058-gd0edd984d
`uname -a`: Linux cooler 5.4.6-arch3-1 #1 SMP PREEMPT Tue, 24 Dec 2019 04:36:53 +0000 x86_64 GNU/Linux
`$TERM`: xterm-256color
Problem occurs in clean fish install: Yes

**Comment by krobelus** [comment]
This is for compatibility with other shells, see #1935. You can use `set -l v; for v in ...` to avoid changing other variables. This happens also for `begin; for v in 1; end` blocks; it's not specific to subshells.

**Comment by zanchey** [comment]
These are not subshells, they are command substitutions.

The function within the command substitution clobbering the local scope outside of the function is probably wrong though.

**Comment by 0ion9** [comment]
@krobelus : isn't #1935 about putting the variables in for (etc) 's parent scope?

In the case with a simple command substitution, the 'parent scope' seems ambiguous to me. What I would have expected is that the command substitution is the parent scope, and hence `v` would be accessible later in the substitution, but not outside of the substitution. Maybe command substitutions don't count as scopes.

Still I'm pretty sure functions normally count as scopes, so the `for` variable being assigned outside of the scope of the function altogether seems unambiguously wrong.

**Comment by krobelus** [comment]
I think command substitutions do have their own local scope, but the `for`-loop sets *global* variable `v`, (unless a local variable `v` exists).
The same problem occurs when sourcing files, e.g. if I have a global/universal variable `x` and then type `cargo `, it will get overwritten by [a loop in cargo.fish](https://github.com/fish-shell/fish-shell/blob/master/share/completions/cargo.fish#L79). (These loops are not necessary but it seems like we should fix this).

So we could interpret
```fish
set i 1
for i in 1 2 3
end
```
as
```fish
set i 1
set -l i $i
for i in 1 2 3
end
```

This means we only have the changed values of the loop variable in the immediate parent of the loop, which is the intention of #1935 as you said. Global and universal variables will never be overwritten by the loop variable.

**Comment by 0ion9** [comment]
That seems much safer and less surprising than the current implementation. I understand it would be slightly less compatible, since other shells have weaker scoping. However, the code that it would break would all be code which is relying on precise  implementation details of a function, which I'm tempted to say is code you really shouldn't expect to work anyway.

**Comment by krobelus** [comment]
I've implemented that now in 992c864.
The function detection was incomplete for command substitutions, that's why it worked for functions but not inside the `()`.
Thanks for tracking this down!

**Comment by zanchey** [comment]
This behaved correctly in 2.7.1, but not 3.0.2, so I think it's reasonable to include this in 3.1.
